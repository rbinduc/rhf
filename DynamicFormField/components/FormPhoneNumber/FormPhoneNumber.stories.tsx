import React from 'react';
import {storiesOf} from '@storybook/react';
import FormPhoneNumber, {IFormPhoneNumberProps} from '.';
import useFormHook from '../../utils/hooks/useFormHook';
import {themeMock} from '../../../../../themes/theme-mock';
import {FormProvider} from 'react-hook-form';
import {ThemeProvider} from 'emotion-theming';

const defaultProps: IFormPhoneNumberProps = {
  id: 'formPhoneNumber',
  name: 'formPhoneNumber',
  label: 'PhoneNumber',
  placeholder: 'Please enter a phone number',
  value: '12345657586',
};

storiesOf('Organisms/RHF/DynamicForm/FormPhoneNumber', module)
  .addDecorator((getStory) => <ThemeProvider theme={themeMock}>{getStory()}</ThemeProvider>)
  .add('default case', () => {
    const {methods}: any = useFormHook();
    return (
      <FormProvider {...methods}>
        <FormPhoneNumber {...defaultProps} />
      </FormProvider>
    );
  });
