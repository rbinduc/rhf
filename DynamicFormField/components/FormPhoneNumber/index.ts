export {default} from './FormPhoneNumber';
export type {IFormPhoneNumberProps} from './FormPhoneNumber';
