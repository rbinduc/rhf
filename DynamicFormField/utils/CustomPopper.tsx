import React from 'react';
import {Popper, PopperProps} from '@material-ui/core';

export const CustomPopper = ({placement = 'bottom', ...rest}: PopperProps) => <Popper {...rest} placement={placement} />;
