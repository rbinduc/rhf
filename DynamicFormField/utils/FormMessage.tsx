import styled from '@emotion/styled';
import {FormHelperText} from '@material-ui/core';
import React from 'react';
import {IconNames} from '../../../../icons/IconNameConstants';
import {TenantVector} from '../../../../icons/TenantVector/TenantVector.component';

interface IMessageProps {
  message?: React.ReactNode;
  isBold?: boolean;
  isLarge?: boolean;
}

export interface IFormMessageProps extends IMessageProps {
  id?: string;
  Icon?: IconNames;
  component?: React.ElementType;
  margin?: 'dense';
  isError?: boolean;
  hideIcon?: boolean;
}

const FormMessage = ({id = '', Icon, message, component, margin, isBold, isLarge, hideIcon, isError = true}: IFormMessageProps) =>
  message ? (
    <FormHelperText id={id} error={!!isError} component={component || 'div'} margin={margin}>
      <FormMessageWrapper>
        {!hideIcon && (
          <span id={id + '_icon'}>
            <TenantVector tenantVectorName={Icon || (isError ? IconNames.FieldError : IconNames.Info)} />
          </span>
        )}
        <Message id={id + '_message'} isBold={!!isBold} isLarge={!!isLarge}>
          {message}
        </Message>
      </FormMessageWrapper>
    </FormHelperText>
  ) : null;

const FormMessageWrapper = styled.span`
  display: flex;
  flex-direction: row;
  & svg {
    margin-top: 0.2rem;
    margin-right: 0.5rem;
  }
  word-break: break-word;
`;

const Message = styled.span<IMessageProps>`
  font-weight: ${({isBold}) => (isBold ? 'bold' : 400)};
  font-size: ${({isLarge}) => (isLarge ? '1rem' : '0.875rem')};
`;

export default FormMessage;
