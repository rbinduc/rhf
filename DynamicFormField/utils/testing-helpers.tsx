import React from 'react';
import {render} from '@testing-library/react';
import {FormProvider} from 'react-hook-form';
import useFormHook from './hooks/useFormHook';
import {ThemeProvider} from 'emotion-theming';
import {themeMock} from '../../../../themes/theme-mock';

interface IDefaultValue {
  [key: string]: any;
}

interface RenderOptions {
  defaultValues?: IDefaultValue;
  toPassBack?: any[];
}

/**
 * Testing Library utility function to wrap tested component in React Hook Form
 * @param {ReactElement} ui A React component
 * @param objectParameters
 * @param {Object} objectParameters.defaultValues Initial form values to pass into
 * React Hook Form, which you can then assert against
 * @param {Array} objectParameters.toPassBack React Hook Form method names which we'd
 * like to pass back and use in tests.  A primary use case is sending back 'setError',
 * so we can manually setErrors on React Hook Form components and test error handling
 */
export function renderWithDyanmicForm(
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  {defaultValues = {}, toPassBack = []}: RenderOptions = {}
) {
  let dynamicFormMethods = {} as any;

  const Wrapper = ({children}: {children: React.ReactNode}) => {
    const {methods}: any = useFormHook({defaultValues});
    for (let dynamicFormItem of toPassBack) {
      dynamicFormMethods[dynamicFormItem] = methods[dynamicFormItem];
    }

    return (
      <ThemeProvider theme={themeMock}>
        <FormProvider {...methods}>{children}</FormProvider>
      </ThemeProvider>
    );
  };

  return {
    ...render(ui, {wrapper: Wrapper as React.ComponentType<{}>}),
    dynamicFormMethods,
  };
}

/**
 * Higher order helper function which wraps a component w/ React Hook Form
 * @param {React Component} WrappedComponent to pass into
 * @param {*} restProps any other remaining props
 * @returns {React Component}
 */
export function withDynamicForm(WrappedComponent: React.ComponentType<any>, restProps: object, defaultValues: IDefaultValue) {
  return () => {
    const methods = useFormHook({defaultValues});

    return (
      <ThemeProvider theme={themeMock}>
        <WrappedComponent {...restProps} {...methods} />
      </ThemeProvider>
    );
  };
}
