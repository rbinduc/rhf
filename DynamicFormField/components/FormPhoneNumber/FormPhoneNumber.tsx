import React from 'react';
import {Box, InputLabel} from '@material-ui/core';
import {Controller, useFormContext, Message} from 'react-hook-form';
import {IDynamicFormFieldCommonProps, getLabel} from '../../DynamicFormField';
import styled from '@emotion/styled';
import {useTheme} from 'emotion-theming';
import {ITenantTheme, IFormElement} from '../../../../../../models';
import PhoneInput from 'react-phone-input-2';
import libphonenumber from 'google-libphonenumber';
import 'react-phone-input-2/lib/high-res.css';
import es from 'react-phone-input-2/lang/es.json';
import fr from 'react-phone-input-2/lang/fr.json';
import ru from 'react-phone-input-2/lang/ru.json';
import FormMessage from '../../utils/FormMessage';

const PLUS_SYMBOL = '+';
const FALLBACK_COUNTRY_CODE = '001';

// ----------------------------------------------------------------------------
// Props
// ----------------------------------------------------------------------------
export interface IFormPhoneNumberProps extends IDynamicFormFieldCommonProps {
  languageCode?: string;
  allowItalianLeadingZero?: boolean;
}

export type IKey = {
  [key: string]: any;
};

const isValidItalianNumber = (phoneNumberInput: libphonenumber.PhoneNumber): boolean => {
  const hasItalianLeadingZero = phoneNumberInput.hasItalianLeadingZero();
  const phoneNumber = phoneNumberInput.getNationalNumber();
  const count = String(phoneNumber)?.length >= 9;

  return !hasItalianLeadingZero && count;
};

// ----------------------------------------------------------------------------
// Component
// ----------------------------------------------------------------------------
const FormPhoneNumber: React.FC<IFormPhoneNumberProps> = (props: IFormPhoneNumberProps) => {
  const {
    validations,
    label,
    name,
    languageCode,
    placeholder,
    isDisabled,
    isRequired,
    defaultValue,
    id,
    labelId = '',
    allowItalianLeadingZero = false,
  } = props;
  const {
    control,
    formState: {errors, dirtyFields},
  } = useFormContext();
  const errorMessage = errors[name]?.message;
  const isDirty = !!dirtyFields[name];
  const phoneUtil = libphonenumber.PhoneNumberUtil.getInstance();
  const theme: ITenantTheme = useTheme();
  const errorClass = !!errorMessage ? 'phone-number-format-error' : '';
  const dirtyClass = isDirty ? 'dirty' : '';
  let containerClass = `phone-number-format ${errorClass} ${dirtyClass}`;

  const rules = {
    ...validations,
    validate: (value: any) => {
      let isValid: Message | boolean = validations?.validateMessage as Message;
      try {
        const inputVal = value?.startsWith(PLUS_SYMBOL) ? value : PLUS_SYMBOL + value;
        const phoneNumberInput = phoneUtil.parseAndKeepRawInput(inputVal);
        const isValidPhoneNumber = phoneUtil.isValidNumber(phoneNumberInput);
        const countryCode = phoneUtil.getRegionCodeForNumber(phoneNumberInput) ?? FALLBACK_COUNTRY_CODE;
        if (isValidPhoneNumber && isNaN(parseInt(countryCode, 10))) {
          isValid = true as boolean;
        }
        if (!allowItalianLeadingZero && countryCode === 'IT') {
          isValid = isValidItalianNumber(phoneNumberInput) || (validations?.validateMessage as Message);
        }
      } catch (e) {}
      return isValid;
    },
  };

  const localeCodes: IKey = {es, fr, ru};
  const lngCode = languageCode?.toLowerCase() ?? 'us';
  const formElement = theme?.designSystem?.atoms?.formElement ?? {};
  return (
    <BoxStyled {...formElement}>
      <InputLabel htmlFor={name} className={errorMessage && 'error'} id={labelId}>
        {getLabel(label, !!isRequired)}
      </InputLabel>
      <Controller
        name={name}
        control={control}
        rules={rules}
        defaultValue={defaultValue}
        render={({field}) => (
          <PhoneInput
            {...field}
            placeholder={placeholder}
            inputProps={{
              id,
              name,
              disabled: isDisabled,
              readOnly: isDisabled,
              required: isRequired,
              'aria-label': label && typeof label === 'string' ? label : placeholder,
              'aria-hidden': isDisabled,
            }}
            containerClass={containerClass}
            localization={localeCodes[lngCode]}
            onChange={(e, country, value, formattedValue) => field.onChange(formattedValue)}
          />
        )}
      />
      <FormMessage message={errorMessage} />
    </BoxStyled>
  );
};

const BoxStyled = styled(({borderColor, borderColorFocused, borderColorError, ...rest}) => <Box {...rest} />)<IFormElement>`
  .phone-number-format {
    display: flex;
    flex-direction: row-reverse;
    justify-content: center;
    border-radius: 0.25rem;

    .form-control {
      height: 3.5rem;
      width: 100%;
      padding-left: 3.5rem;
      padding: 1.15625rem 0.875rem;
      margin-left: 0.5rem;
      border-radius: inherit;
      border-color: ${({borderColor}) => borderColor};
      & + .flag-dropdown {
        border-color: ${({borderColor}: IFormElement) => borderColor};
      }
    }
    .form-control:focus,
    &.dirty .form-control {
      border-width: 0.125rem;
      border-color: ${({borderColorFocused}: IFormElement) => borderColorFocused};
      & + .flag-dropdown {
        border-width: 0.125rem;
        border-color: ${({borderColorFocused}: IFormElement) => borderColorFocused};
      }
    }

    .flag-dropdown {
      border-radius: inherit;
      position: initial;
      background-color: transparent;
      .country-list {
        width: 100%;
        border: ${({borderColorFocused}: IFormElement) => `1px solid ${borderColorFocused}`};
        border-radius: 0.3125rem;
        box-shadow: 0rem 0.125rem 0.375rem #dbdbdb;
        .country {
          padding: 0.625rem 0.5625rem 0.625rem 3.125rem;
          position: relative;
          height: 2.5rem;
          border-bottom: 0.0625rem solid #dbdbdb;
          line-height: 1.4rem;
          font-weight: 400;
          font-size: 1rem;
          font-family: Nunito;
          display: flex;
          flex-direction: row;
          &:hover,
          &.highlight {
            background-color: #c7ced6;
          }
          .country-name {
            max-width: 200px;
            text-overflow: ellipsis;
            white-space: nowrap;
            overflow: hidden;
          }
        }
        .flag {
          left: 0.9375rem;
          top: 0.5rem;
          transform: scale(1.2);
        }
      }
    }

    .flag-dropdown.open .selected-flag {
      border-radius: inherit;
    }

    .selected-flag {
      width: 5.25rem;
      border-radius: inherit;
      padding: 0;
      display: flex;
      justify-content: center;
      .flag {
        margin-left: -0.9375rem;
        transform: scale(1.2);
      }
      [class='flag 0'] {
        margin-top: -0.625rem;
        background-image: url("data:image/svg+xml,%3Csvg width='24' height='24' viewBox='0 0 24 24' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M15.9062 15.3906C17.3125 13.8906 18.0156 12.0938 18.0156 10C18.0156 8.34375 17.5469 6.84375 16.6094 5.5C15.7031 4.15625 14.5 3.1875 13 2.59375V3.01562C13 3.54688 12.7969 4.01562 12.3906 4.42188C11.9844 4.79688 11.5156 4.98438 10.9844 4.98438H9.01562V7C9.01562 7.28125 8.90625 7.51562 8.6875 7.70312C8.5 7.89062 8.26562 7.98438 7.98438 7.98438H6.01562V10H12.0156C12.2969 10 12.5312 10.0938 12.7188 10.2812C12.9062 10.4688 13 10.7031 13 10.9844V13.9844H13.9844C14.9219 13.9844 15.5625 14.4531 15.9062 15.3906ZM9.01562 17.9219V16C8.48438 16 8.01562 15.7969 7.60938 15.3906C7.20312 14.9844 7 14.5156 7 13.9844V13L2.21875 8.21875C2.0625 8.84375 1.98438 9.4375 1.98438 10C1.98438 12.0312 2.65625 13.7969 4 15.2969C5.375 16.7969 7.04688 17.6719 9.01562 17.9219ZM2.92188 2.96875C4.89062 1 7.25 0.015625 10 0.015625C12.75 0.015625 15.0938 1 17.0312 2.96875C19 4.90625 19.9844 7.25 19.9844 10C19.9844 12.75 19 15.1094 17.0312 17.0781C15.0938 19.0156 12.75 19.9844 10 19.9844C7.25 19.9844 4.89062 19.0156 2.92188 17.0781C0.984375 15.1094 0.015625 12.75 0.015625 10C0.015625 7.25 0.984375 4.90625 2.92188 2.96875Z' fill='%23273446'%3E%3C/path%3E%3C/svg%3E");
      }

      .arrow {
        border-left-width: 0.25rem;
        border-right-width: 0.25rem;
      }
    }
  }
  .phone-number-format-error,
  .phone-number-format-error.dirty {
    & .form-control {
      border-width: 0.125rem;
      border-color: ${({borderColorError}: IFormElement) => borderColorError};
      & + .flag-dropdown {
        border-width: 0.125rem;
        border-color: ${({borderColorError}: IFormElement) => borderColorError};
      }
      &:focus {
        border-color: ${({borderColorError}: IFormElement) => borderColorError};
        & + .flag-dropdown {
          border-color: ${({borderColorError}: IFormElement) => borderColorError};
        }
      }
    }
  }
`;

export default FormPhoneNumber;
