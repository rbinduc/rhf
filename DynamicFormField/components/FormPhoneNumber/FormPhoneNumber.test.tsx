import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import {screen, act} from '@testing-library/react';
import FormPhoneNumber, {IFormPhoneNumberProps} from './FormPhoneNumber';
import {renderWithDyanmicForm} from '../../utils/testing-helpers';
import {UseFormReturn} from 'react-hook-form';

const formPhoneNumberProps: IFormPhoneNumberProps = {
  id: 'FormPhoneNumber',
  name: 'FormPhoneNumber',
  label: 'Form PhoneNumber',
  placeholder: 'E.g.1234567890',
  defaultValue: '+1 (234) 565-7586',
};

jest.mock('@ionic-native/keyboard', () => ({
  hide: jest.fn(),
}));

jest.mock('../../../../../../automation/PageLocators', () => ({
  locators: {
    jobPageLocators: {
      aFiltersIcon: 'filtersIcon',
    },
    commonLocators: {
      aHeaderText: 'headerText',
    },
    accountSettings: {
      aChangeCountryHeader: 'changeCountryHeader',
    },
  },
}));

describe('FormPhoneNumber', () => {
  it('snapshot match', () => {
    const {container} = renderWithDyanmicForm(<FormPhoneNumber {...formPhoneNumberProps} />, {});
    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();

    expect(screen).toMatchSnapshot();
  });

  it('renders in document and a default value and disabled', () => {
    renderWithDyanmicForm(<FormPhoneNumber {...formPhoneNumberProps} isDisabled={true} />, {});
    const textbox = screen.getByLabelText(formPhoneNumberProps.label as string);

    expect((textbox as HTMLInputElement).disabled).toBeTruthy();
    expect((textbox as HTMLInputElement).value).toEqual(formPhoneNumberProps.defaultValue);
  });

  it('textbox aria role is applied once (implicitly for input)', () => {
    renderWithDyanmicForm(<FormPhoneNumber {...formPhoneNumberProps} />, {});
    expect(screen.getAllByRole('textbox').length).toEqual(1);
  });

  it('displays error message when errors', async () => {
    const {getByText, dynamicFormMethods} = renderWithDyanmicForm(<FormPhoneNumber {...formPhoneNumberProps} />, {
      toPassBack: ['setError'],
    });

    await act(async () => {
      (dynamicFormMethods as UseFormReturn)?.setError(formPhoneNumberProps.name, {
        type: 'manual',
        message: 'PhoneNumber is required',
      });
    });

    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(getByText('PhoneNumber is required'));
  });
});
