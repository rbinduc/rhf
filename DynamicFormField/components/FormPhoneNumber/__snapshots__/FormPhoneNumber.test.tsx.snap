// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormPhoneNumber snapshot match 1`] = `
<div
  class="MuiBox-root MuiBox-root-1 css-1e0xmn4-BoxStyled e1ezzjsz0"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
    for="FormPhoneNumber"
    id=""
  >
    Form PhoneNumber
  </label>
  <div
    class="phone-number-format   react-tel-input"
  >
    <div
      class="special-label"
    >
      Phone
    </div>
    <input
      aria-label="Form PhoneNumber"
      class=" form-control"
      id="FormPhoneNumber"
      name="FormPhoneNumber"
      placeholder="E.g.1234567890"
      type="tel"
      value="+1 (234) 565-7586"
    />
    <div
      class=" flag-dropdown"
    >
      <div
        aria-haspopup="listbox"
        class="selected-flag"
        role="button"
        tabindex="0"
        title="United States: + 1"
      >
        <div
          class="flag us"
        >
          <div
            class="arrow"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormPhoneNumber snapshot match 2`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
